### PR TITLE
Manage RaftNodeAsClient in cluster 

### PR DESCRIPTION
**Pull request by fanhualta**

### CODE REVIEW

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 1]
```diff
@@ -62,81 +64,139 @@
...
-   * Lock to update validClientNum
+   * Lock to update clientNumInUse
    */
-  private ReentrantLock numLock = new ReentrantLock();
+  private ReentrantLock resourceLock = new ReentrantLock();
+
+  /**
+   * Mark whether system is shutting down
+   */
+  private AtomicBoolean isShuttingDown = new AtomicBoolean(false);
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: add volatile is better?

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 2]
```diff
@@ -62,81 +64,139 @@
...
+  private RaftNodeAsClient tryToGetClient() throws RaftConnectionException {
     for(;;){
-      if(validClientNum.get() < MAX_VALID_CLIENT_NUM){
+      if(clientNumInUse.get() < MAX_VALID_CLIENT_NUM){
+        resourceLock.lock();
         try{
-          numLock.lock();
-          if(validClientNum.get() < MAX_VALID_CLIENT_NUM){
-            validClientNum.incrementAndGet();
+          checkShuttingDown();
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: If queue is full, try sleep(1000) is better? `for(;;)` wastes cpu time

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 3]
```diff
@@ -62,81 +64,139 @@
...
+    } finally {
+      resourceLock.unlock();
+    }
   }
 
   public void shutdown(){
-    client.shutdown();
+    isShuttingDown.set(true);
+    while (clientNumInUse.get() != 0 && queueClientNum != 0){
+      // wait until releasing all usage of clients.
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: try sleep some time?

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 4]
```diff
@@ -171,14 +231,13 @@ private RaftNodeAsClient(){
...
       boltClientService.init(new CliOptions());
-      clientInited = true;
     }
 
     @Override
     public void asyncHandleRequest(BasicRequest request, PeerId leader,
         QPTask qpTask)
         throws RaftConnectionException {
-      LOGGER.debug("Node as client to send request to leader: {}", leader);
+      LOGGER.debug("Node as clientList to send request to leader: {}", leader);
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: Node as client?

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 5]
```diff
@@ -187,15 +246,15 @@ public void asyncHandleRequest(BasicRequest request, PeerId leader,
...
                     BasicResponse response = (BasicResponse) result;
-                    releaseClient();
+                    releaseClient(RaftNodeAsClient.this);
                     qpTask.run(response);
                   }
 
                   @Override
                   public void onException(Throwable e) {
-                    LOGGER.error("Bolt rpc client occurs errors when handling Request", e);
+                    LOGGER.error("Bolt rpc clientList occurs errors when handling Request", e);
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: Bolt rpc client?

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 6]
```diff
@@ -207,7 +266,7 @@ public Executor getExecutor() {
       } catch (RemotingException | InterruptedException e) {
         LOGGER.error(e.getMessage());
         qpTask.setTaskState(TaskState.EXCEPTION);
-        releaseClient();
+        releaseClient(this);
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: why not RaftNodeAsClient.this?

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java [hunk 7]
```diff
@@ -226,17 +285,16 @@ public void syncHandleRequest(BasicRequest request, PeerId leader,
         qpTask.run(null);
         throw new RaftConnectionException(e);
       } finally {
-        releaseClient();
+        releaseClient(this);
```
Comment (MyXOF) [inline cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: why not RaftNodeAsClient.this?

Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: It's the same, but pointing it out clearly is better.

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java (hunk 6 above)
Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: It's the same, but pointing it out clearly is better.

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java (hunk 5 above)
Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: fixed

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java (hunk 4 above)
Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: fixed

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java (hunk 3 above)
Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: fixed

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java (hunk 1 above)
Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: yes, accept it.

File: cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java (hunk 2 above)
Comment (fanhualta) [reply cluster/src/main/java/org/apache/iotdb/cluster/rpc/raft/impl/RaftNodeAsClientManager.java]: fixed

**Pull request merged**